File: src/main/java/no/fintlabs/request/RequestFintEventEntity.java
package no.fintlabs.request;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import lombok.*;

@Entity
@Getter
@NoArgsConstructor
public class RequestFintEventEntity{

    @Id
    public String corrId;
    public String orgId;
    public String topic;
    public String domainName;
    public String packageName;
    public String resourceName;
    public String operationType;
    public long created;
    public long timeToLive;
    public String value;

    public RequestFintEventEntity(String corrId, String orgId, String topic, String domainName, String packageName, String resourceName, String operationType, long created, long timeToLive, String value) {
        this.corrId = corrId;
        this.orgId = orgId;
        this.topic = topic;
        this.domainName = domainName;
        this.packageName = packageName;
        this.resourceName = resourceName;
        this.operationType = operationType;
        this.created = created;
        this.timeToLive = timeToLive;
        this.value = value;
    }
}

File: src/main/kotlin/no/fintlabs/event/request/RequestFintEventEntity.kt
package no.fintlabs.event.request

import jakarta.persistence.Entity
import jakarta.persistence.Id
import no.fintlabs.adapter.models.event.RequestFintEvent
import no.fintlabs.adapter.operation.OperationType

@Entity
class RequestFintEventEntity(
    @Id var corrId: String,
    var orgId: String,
    var topic: String,
    var domainName: String,
    var packageName: String,
    var resourceName: String,
    var operationType: String,
    var created: Long,
    var timeToLive: Long,
    var value: String
) {
    constructor() : this("", "", "", "", "", "", "", 0L, 0L, "")

    fun toRequestEvent(): RequestFintEvent {
        return RequestFintEvent.builder()
            .corrId(corrId)
            .orgId(orgId)
            .domainName(domainName)
            .packageName(packageName)
            .resourceName(resourceName)
            .operationType(parseOperationType(operationType))
            .created(created)
            .timeToLive(timeToLive)
            .value(value)
            .build()
    }

    private fun parseOperationType(operationTypeValue: String?): OperationType {
        return when (operationTypeValue) {
            "CREATE" -> OperationType.CREATE
            "UPDATE" -> OperationType.UPDATE
            "VALIDATE" -> OperationType.VALIDATE
            else -> OperationType.UNKNOWN
        }
    }

    companion object {
        fun fromRequestEvent(requestEvent: RequestFintEvent, topic: String): RequestFintEventEntity {
            return RequestFintEventEntity(
                requestEvent.corrId,
                requestEvent.orgId,
                topic,
                requestEvent.domainName,
                requestEvent.packageName,
                requestEvent.resourceName,
                requestEvent.operationType.name,
                requestEvent.created,
                requestEvent.timeToLive,
                ""
            )
        }
    }
}